### ISSUE TITLE
Resampling weight type in drizpars file is ignored

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#7867]
_Issue [JP-3360](https://jira.stsci.edu/browse/JP-3360) was created on JIRA by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke):_

While characterizing NIRSpec calibration for MOS data, we noticed that rectified images had much better data quality for the bright calibration sources when using weight_type = 'exptime' for the ResampleSpec step, rather than the default inverse variance weighting ('ivm'). We can set the weight_type to 'exptime' directly, with e.g.:

> `s2d = ResampleSpecStep.call(f, weight_type='exptime')`

We are considering doing more testing to decide if we should modify the default parameter value for weight_type, but there appears to be a bug in the resample code making it impossible to change the default value via reference file.  The weight type in the reference file is only used if you explicitly set weight_type=None in the parameter steps, e.g.:

> `s2d = ResampleSpecStep.call(f, weight_type=None, override_drizpars='jwst_nirspec_drizpars_exptime.fits')`

will use weight_type = 'exptime'. Otherwise, it ignores the value in the reference file:

> `s2d = ResampleSpecStep.call(f, override_drizpars='jwst_nirspec_drizpars_exptime.fits')`

will use weight_type='ivm'.

We'd also like to test unweighted resampling (weight_type=None), but this is not possible to set manually:

> `s2d = ResampleSpecStep.call(f, weight_type=None)`

will use weight_type='ivm'.   It is possible via reference file, with weight_type=None also specified. The pipeline issues a validation error because it expects only 'ivm' or 'exptime', but it performs unweighted resampling anyway.

We'd like these interface issues cleaned up so that we can test the three implemented weighting methods and possibly set a different default value.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=659319#comment-659319):_

Attached sample drizpars reference files with the three weight_type options.

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=659706#comment-659706):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) when you get a chance, could you add a criticality/rationale to this ticket? Just so we don't accidentally miss it when searching for high priority tickets

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=660476#comment-660476):_

Tagging [Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra) as there has been a discussion within MIRI about a very similar topic that should get cross-linked to this one.

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=673736#comment-673736):_

[Misty Cracraft](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cracraft) [Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra) I know Mattia is leaving, do either of you know the ticket David is referring to? I'd like to make sure the issues are cross-linked.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=673792#comment-673792):_

[Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra) I couldn't remember if there was a ticket either.  Assuming there isn't one, could you summarize what you've been seeing briefly in either a JP or a JWST-MIRI ticket?  I'd like to make sure we don't lose track of that work.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=673807#comment-673807):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) let me retrieve some plots and I will open an improvement JP ticket (because I believe the code does what it is supposed to do, it is just a matter of choosing the best set of parameters).

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=675674#comment-675674):_

The problem here is the fact that the default value for `weight_type` is set to "ivm" in the resample_step.py module. That makes it look identical to a user-provided value, which always takes precedence over values from a ref file. So simply resetting the default value of `weight_type` to `None` in the step code will then allow the value to be taken from the drizpars ref file.

Currently the drizpars ref files in CRDS for MIRI, NIRCam, and NIRISS all have `weight_type` set to "exptime". So making the change to the code mentioned above would then cause all use of resample by those instruments to switch to using "exptime" weighting, as specified in the drizpars ref file. NIRSpec use of resample_spec could then also be changed to use "exptime" if they were to submit a drizpars ref file to CRDS that has `weight_type` set that way.

If not all instruments want to switch to using "exptime", they could submit updated drizpars ref files with it set to "ivm".

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=675846#comment-675846):_

That sounds like a reasonable resolution to me - if there are current reference files, they should be used by the pipeline or else removed.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=679508#comment-679508):_

[Rosa Diaz](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rdiaz) [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) Can we bring this up in a DMS WG meeting, in order to get consensus on how to proceed and get it done? The proposal is to remove the default setting of `weight_type = ivm` in the resample step code and instead have the default in the code be "None", so that the setting of `weight_type = exptime` in all of the current drizpars reference files gets used. Note that this means **ALL** resampling for all instrument modes will switch from ivm to exptime weighting.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=685018#comment-685018):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) I've done some testing on the potential use of either a drizpars ref file or pars-resamplespecstep ref file for NIRSpec, in order to switch over to "exptime" weighting without the need for step code updates. As we already knew, delivering a NIRSpec drizpars ref file to CRDS with weight type set to "exptime" won't result in any actual changes in step behavior unless we also modify the step code to remove the hardwired default param settings (but that would also suddenly cause all the other instrument's drizpars ref files to become active and result in changes too, which may not be desired).

So the other possible route is the implementation of a pars-resamplespecstep ref file. This could be done easily, because that ref file type is already defined in CRDS, it just doesn't have any actual ref files installed. The good news is that my testing shows that this approach will likely work. I created a pars-resamplespecstep ref file and fed it directly to the resample_spec step when running it on a random NIRSpec fixed-slit "cal" file and it did use the non-default param values (i.e. weight_type = 'exptime') that I specified in the pars ref file. It still issues a warning that no drizpars ref file was found in CRDS for NIRSpec and jumps to the branch in the code that sets default values to use, but that code has appropriate logic in it to recognize when param values have already been specified upstream and hence not reset them to hardwired default values.

Attached is the example pars ref file that I used for testing. You could use that as a template for creating your own custom version.

The way I tested it was by executing (from the command line):
```java
strun pars-resamplespecstep.asdf <random_nrs_cal.fits> --disable-crds-steppars```
Note that the use of the `--disable-crds-steppars` param in this case is just because I don't want it to go looking for a pars ref file in CRDS, because there isn't one (yet) and when it doesn't find one in CRDS it may resort back to the usual defaults that we don't want. This will not be necessary once your pars ref file gets delivered to CRDS.

So, bottom line, it looks like the delivery of a pars ref file can be used to result in changing the weight_type for NIRSpec, without the need for any code changes.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=685023#comment-685023):_

Thanks so much for testing that, [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse), and for the example file.  We'll look into delivering the override for NIRSpec as soon as we can.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=685030#comment-685030):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) Regarding the example pars ref file, I generated it by asking strun to save params for me when I ran the step, which naturally results in it saving ALL params to the file. I then only edited the weight_type param in the resulting file (and other misc stuff, like author, useafter, etc. just so they didn't have blank values). But for the step params themselves, you don't need all of them to be listed in the file that you deliver to CRDS - only the ones for which you want to use a non-default value. In fact it's not a good idea to include all params by default, because if any of the step params should change in the future (e.g. an old param gets removed or a param gets renamed) the ref file will then fail validation when it's loaded into the pipeline. So best practice is to only include the ones you need to change values. (but maybe you already knew this ...)

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=685034#comment-685034):_

Thanks again! That was what I expected, and I am experimenting locally with taking out all the parameters except the ones I want to modify, but it's good to have confirmation that that's the right thing to do.

**Comment by stscijgbot-jp** [comment]
_Comment by [Cheryl Pavlovsky](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cherylp) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=685043#comment-685043):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) The pars file you attached does not pass certification because it does not have a meta history section. What is, or where do I find, the formatting for that?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=685182#comment-685182):_

[Cheryl Pavlovsky](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cherylp) Try this updated version, which now has a history entry.

[^pars-resamplespecstep_history.asdf]

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=702525#comment-702525):_

JP-2682 replaces the use of the drizpars ref file with resample/resample_spec parameter ref files. Once those pars ref files become active in a CRDS context, the drizpars ref file will be obsolete and irrelevant, and this issue will be declared moot. Will close this ticket at that time.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3360?focusedCommentId=703658#comment-703658):_

Closing this ticket now that the use of the drizpars ref file has been removed by JP-2682.
